server/src/main/java/com/cloud/api/query/QueryManagerImpl.java

```diff
@@ -3758,6 +3758,9 @@ else if (!template.isPublicTemplate() && caller.getType() != Account.Type.ADMIN)
                 }
             } else if (templateFilter == TemplateFilter.sharedexecutable || templateFilter == TemplateFilter.shared) {
                 // only show templates shared by others
+                if (permittedAccounts.isEmpty()) {
+                    return new Pair<>(new ArrayList<>(), 0);
+                }
                 sc.addAnd("sharedAccountId", SearchCriteria.Op.IN, permittedAccountIds.toArray());
```

Comment:
seems a bit like a hack on a hack, shouldn´t this be handled by the `SearchCriteria<>.addAnd()`?

Reply:
@DaanHoogland 
no idea. I just verified this change fixes the issue.
if you find another solution, it would be better, I will close this PR in that case.

Reply:
I researched a little and the handling of an empty set is going to be just as ugly. Let's go with this.